### DISCUSSION TITLE
Latest three posts for each blog

### DISCUSSION THREAD

**Discussion by NickDarvey** [OrchardCMS/OrchardCore#8657] [Q&A] [answered]
I have something similar to the TheBlogTheme where there's collections (e.g. Blogs) which contain items (e.g. BlogPosts). I'd like to have a page where I display the latest three blog posts from each blog, grouped by the blog (and displaying the blog title, icon).

Right now I'm using Orchard in a decoupled way, so I'm particularly looking for what kinds of querying I should be doing with something like `IOrchardHelper` to fetch the results rather than displaying them (and without a ridiculous number of queries).

I know I have lots of options open to me with Lucene, YesSQL queries, etc... but I'm not quite sure where to begin!

**Comment by Skrypt** [comment, marked as answer]
I would have a tendency to try with Lucene Queries for that one. Though a SQL Query would definitely work too.

SQL Query (UNION can be slow) : 

```SQL
SELECT DocumentId FROM (
SELECT TOP 3 DocumentId, CreatedUtc FROM ContentItemIndex WHERE ContentType = 'ContentTypeName1' AND Published = 1 AND Latest = 1
UNION
SELECT TOP 3 DocumentId, CreatedUtc FROM ContentItemIndex WHERE ContentType = 'ContentTypeName2' AND Published = 1 AND Latest = 1
UNION
SELECT TOP 3 DocumentId, CreatedUtc FROM ContentItemIndex WHERE ContentType = 'ContentTypeName3' AND Published = 1 AND Latest = 1
ORDER BY CreatedUtc DESC
) as TempTable
```

From this Query you would check the option "Return content items" from the OC SQL Query.

Lucene Query (faster) : 

```json
{
  "from" : {{ from }}, "size" : {{ size }},
  "query": {
    "bool" : {
      "must" : [
        { "term" : { "Content.ContentItem.Published" : "true" }},
        {
          "bool": {
            "should": [
              { "term" : { "Content.ContentItem.ContentType" : { "value": "AccessoryEquipment", "boost": 1 }}}
            ]
          }
        }
      ]
    }
  },
  "sort": [
    { "Content.ContentItem.CreatedUtc": { "order": "desc" } }
  ]
}
```

Here with Lucene I think you would require 3 different Queries (one per content type) because of the "size" which can only be done per Query.

Need to see if we could use a MultiSearch query in OC.
https://www.elastic.co/guide/en/elasticsearch/reference/current/search-multi-search.html#search-multi-search

And of course in your Razor view you need to use something like (after injecting IOrchardHelper in your _ViewImport.cshtml file) : 

```C#
var someCollection = (IEnumerable<ContentItem>)await Orchard.QueryAsync("YourQueryName");
```

**Reply by NickDarvey** [reply]
Hmm, not quite what I was thinking. I don't think I articulated it well, but there's only two content types involved: 'Blog' and 'BlogPost'.

Perhaps I can express the problem another way. If I were trying to express this querying a relational database it would be something like:

```sql
select b.BlogName
     , p.Title
     , p.CreatedAt
from Blogs as b
inner join (
  select top(3) Title, CreatedAt, BlogId
  from Posts
  order by CreatedAt desc
  ) p
  on p.BlogId = b.Id
```
(I'm not sure if my SQL is correct... but hopefully you get the idea.)

Also, it would be brilliant to get the results back in as content, i.e. something like:
```json
{
  "results": [
    {
      "id": "1",
      "blogName": "My First Blog",
      "posts": [
        {
          "id": "A",
          "title": "A Post",
          "createdAt": "2021-02-23T00:00Z"
        }
        {
          "id": "B",
          "title": "B Post",
          "createdAt": "2021-01-23T00:00Z"
        }
        {
          "id": "C",
          "title": "C Post",
          "createdAt": "2020-12-23T00:00Z"
        }
      ]
    },
    {
      "id": "2",
      "blogName": "My Second Blog",
      "posts": [
        {
          "id": "D",
          "title": "D Post",
          "createdAt": "2021-02-19T00:00Z"
        }
        {
          "id": "E",
          "title": "E Post",
          "createdAt": "2021-01-18T00:00Z"
        }
        {
          "id": "F",
          "title": "F Post",
          "createdAt": "2020-12-17T00:00Z"
        }
      ]
    }
  ]
}

```

**Reply by Skrypt** [reply]
Okay but I should have asked if you are trying to display these from actual Content Types like Blog and BlogPost or is it coming from a custom table you made in your database?

I see your point there. This JSON is clean and relevant for your use case. If you want to receive only small amount of JSON data you can use the GraphQL endpoint to retrieve these. But you need to have a SQL or Lucene Query mapped in the GraphQL endpoint first. You can use the Schema part of the Query to define these fields but they won't render exactly as the JSON you have shown there.

If you want to reproduce this exact JSON structure I think right now you will need to do some custom API endpoint or add it by yourself to the GraphQL endpoint.

**Reply by NickDarvey** [reply]
These are actual Content Types like Blog and BlogPost.

Is this the kind of query you can write in Lucene?

It doesn't need to be exactly like the JSON I showed, so long as the content is getting returned :)

**Reply by Skrypt** [reply]
Yes, there is a checkbox on our OC Queries to return content items complete objects. But these contains all the JSON we are storing in the Document table in the Content column. So, if you look at a "BlogPost" content data you will see that we are not storing the "Blog" data in them but a reference of the BlogPost to it's ContainedPart.

```json
{
  "ContentItemId": "4cr3mj25j0qwb6txxx8wznxfnq",
  "ContentItemVersionId": "4pshqt4kbje85v6asshkmeagw1",
  "ContentType": "BlogPost",
  "DisplayText": "Man must explore, and this is exploration at its greatest",
  "Latest": true,
  "Published": true,
  "ModifiedUtc": "2021-02-22T20:16:42.4401381Z",
  "PublishedUtc": "2021-02-19T17:24:22.0679312Z",
  "CreatedUtc": "2021-02-19T17:24:22.0411505Z",
  "Owner": "4shv08fs7smrft92cfspkg76w1",
  "Author": "admin",
  "TitlePart": {
    "Title": "Man must explore, and this is exploration at its greatest"
  },
  "ContainedPart": {
    "ListContentItemId": "4dftc2g2s4t6mtn4e16w0ajg4y",
    "Order": 0
  },
  "MarkdownBodyPart": {
    "Markdown": "Never in all their history have men been able truly to conceive of the world as one: a single sphere, a globe, having the qualities of a globe, a round earth in which all the directions eventually meet, in which there is no center because every point, or none, is center — an equal earth which all men occupy as equals. The airman's earth, if free men make it, will be truly round: a globe in practice, not in theory.\n\nScience cuts two ways, of course; its products can be used for both good and evil. But there's no turning back from science. The early warnings about technological dangers also come from science.\n\nWhat was most significant about the lunar voyage was not that man set foot on the Moon but that they set eye on the earth.\n\nA Chinese tale tells of some men sent to harm a young girl who, upon seeing her beauty, become her protectors rather than her violators. That's how I felt seeing the Earth for the first time. I could not help but love and cherish her.\n\nFor those who have seen the Earth from space, and for the hundreds and perhaps thousands more who will, the experience most certainly changes your perspective. The things that we share in our world are far more valuable than those which divide us."
  },
  "AutoroutePart": {
    "Path": "blog/post-1"
  },
  "BlogPost": {
    "Subtitle": {
      "Text": "Problems look mighty small from 150 miles up"
    },
    "Image": {
      "Paths": [
        "post-bg.jpg"
      ],
      "MediaTexts": [
        ""
      ],
      "Anchors": [
        {
          "X": 0.5,
          "Y": 0.5
        }
      ]
    },
    "Tags": {
      "TermContentItemIds": [
        "4xa02qtfgdfn11gmgrq95np0cq",
        "41s9z8zrcxacf0w1b03k6z0a50",
        "4acsa4ycwj1870nee18n7t6vdx"
      ],
      "TaxonomyContentItemId": "47gja3jr3whs568xbeh9rxsxcq",
      "TagNames": [
        "Earth",
        "Exploration",
        "Space"
      ]
    },
    "Category": {
      "TermContentItemIds": [
        "4pagb7xmct4zp2b3pa670anhmk"
      ],
      "TaxonomyContentItemId": "40d0hnrn8rkfb42786eerc84wr"
    }
  }
}
```

So, yes it requires to do Lazy Loading of the related Content Items by using the different UUID's shown in this JSON.

**Reply by lampersky** [reply]
@NickDarvey you could also use this query (I tried with Blog recipe)

```
select cii.ContentItemId as ListId, cii.DisplayText as ListName,
post.ContentItemId as PostId, post.PublishedUtc
from ContainedPartIndex cpi 
join ContentItemIndex cii on cpi.ListContentItemId = cii.ContentItemId
join ContentItemIndex post on post.DocumentId = cpi.DocumentId
where 
(
  select count(1)
  from ContainedPartIndex cpi2 
  join ContentItemIndex cii2 on cpi2.ListContentItemId = cii2.ContentItemId
  join ContentItemIndex post2 on post2.DocumentId = cpi2.DocumentId
  where 
  cii2.ContentItemId = cii.ContentItemId and
  post2.PublishedUtc >= post.PublishedUtc
) <= 3;
```

Your problem could be described as "how to list N elements per each group", which can be solved in many ways using plain SQL.
I've choosen correlated subquery which should works on all databases.

![image](https://user-images.githubusercontent.com/6403130/109384582-1e401d80-78ee-11eb-9d60-431709bde25f.png)

**Reply by Skrypt** [reply]
Yeah this is another one that could work. Though, you are not returning the Document.Id only here so that when you check on the "return content items" checkbox of the Query that it does so. Here, the Query will return a json object result.
